File: DeleteHalfNodes.py

class Node:
    def __init__(self, x):
        self.data = x
        self.left = None
        self.right = None

class Tree:
    def __init__(self):

        self.root = None
    
    def add(self, x):

        if self.root == None:

            self.root = Node(x)
            return

        else:
            prev = None
            temp = self.root

            while temp != None:
                if x > temp.data:
                    prev = temp
                    temp = temp.right
                
                elif x < temp.data:
                    prev = temp
                    temp = temp.left
            
            if x > prev.data:
                new = Node(x)
                prev.right = new
            
            else:
                new = Node(x)
                prev.left = new
        
        return
    
    def inorderUtil(self, root):
        if root == None:
            return 
        else:
            self.inorderUtil(root.left)
            print(root.data)
            self.inorderUtil(root.right)
        return

    def inorder(self):
        temp = self.root
        self.inorderUtil(temp)
        return

    def PreorderUtil(self, root):
        if root == None:
            return 
        else:
            print(root.data)
            self.PreorderUtil(root.left)
            self.PreorderUtil(root.right)
        return
        
    def Preorder(self):
        temp = self.root
        self.PreorderUtil(temp)
        return

    def PostorderUtil(self, root):
        if root == None:
            return 
        else:
            self.PostorderUtil(root.left)
            self.PostorderUtil(root.right)
            print(root.data)
        return
        
    def Postorder(self):
        temp = self.root
        self.PostorderUtil(temp)
        return


def deleteHalfNodesUtil(tree, prev, connect, root):
    if (root.left== None and root.right == None):
        return

    if (root.left == None and root.right != None):
        if prev == None:
            tree.root = root.right
            deleteHalfNodesUtil(tree, None, 'left', root.right)

        elif connect == 'left':
            prev.left = root.right
            deleteHalfNodesUtil(tree, prev, 'left', root.right)
        else:
            prev.right = root.right
            deleteHalfNodesUtil(tree, prev, 'right', root.right)

    elif (root.left != None and root.right == None):
        if prev == None:
            tree.root = root.left
            deleteHalfNodesUtil(tree, None, 'left', root.left)
            
        elif connect == 'left':
            prev.left = root.left
            deleteHalfNodesUtil(tree, prev, 'left', root.left)
        else:
            prev.right = root.left
            deleteHalfNodesUtil(tree, prev, 'right', root.left)
    
    else:
        deleteHalfNodesUtil(tree, root, 'left', root.left)
        deleteHalfNodesUtil(tree, root, 'right', root.right)
        return


def deleteHalfNodes(tree):
    prev = None
    connect = 'left'
    root = tree.root
    deleteHalfNodesUtil(tree, prev, connect, root)


tree = Tree()

tree.add(3)
tree.add(5)
tree.add(6)
tree.add(4)
tree.add(7)
tree.inorder()
tree.Preorder()
tree.Postorder()
deleteHalfNodes(tree)
tree.inorder()

File: C++/DeleteHalfNodes.cpp
#include <iostream>
using namespace std;

class Tree{
    private:
    struct node{
        int data;
        node* left;
        node* right;
    };
    node* root;

    void add(int x, node** p){

        if( *p == NULL){

            node* newnode = new node;
            newnode->data = x;
            newnode->left = NULL;
            newnode->right = NULL;
            *p = newnode;

            return;
        }
        else{
            if (x > (*p)->data){
                add(x, &(*p)->right);
            }
            else{
                add(x, &(*p)->left);
            }
        }

        return;
    }

    void inorder(node* root){

        if(root==NULL) return;
        
        else{
            inorder(root->left);
            cout<<root->data<<'\n';
            inorder(root->right);
        }

        return;
    }

    void postorder(node* root){

        if(root == NULL) return;

        else{
            postorder(root->left);
            postorder(root->right);
            cout<<root->data<<"\n";
        }
    }

    void preorder(node* root){

        if(root == NULL) return;

        else{
            cout<<root->data<<"\n";
            preorder(root->left);
            preorder(root->right);
        }
    }

    public:
    Tree(){
        this->root = NULL;
    }

    void add(int x){

        node** p = &(this->root);
        this->add(x, p);
    }

    void inorder(){
        inorder(this->root);
    }

    void preorder(){
        preorder(this->root);
    }

    void postorder(){
        postorder(this->root);
    }

    void DeleteHalfNodes(node** temp){
        if (*temp == NULL) {
            return;
        }

        else if((*temp)->right == NULL && (*temp)->left == NULL)  return;

        else if((*temp)->right != NULL && (*temp)->left == NULL){
            *temp = (*temp)->right;
            DeleteHalfNodes(temp);
        }
        else if ((*temp)->right == NULL && (*temp)->left != NULL){
            *temp = (*temp)->left;
            DeleteHalfNodes(temp);
        }
        else{
            DeleteHalfNodes(&(*temp)->left);
            DeleteHalfNodes(&(*temp)->right);
        }
        return;
    }

    void DeleteHalfNodes(){
        node** temp = &root;

        DeleteHalfNodes(temp);
    }

};


int main(){

    Tree tree;
    tree.add(5);
    tree.add(1);
    tree.add(3);
    tree.add(2);
    tree.add(4);
    tree.add(6);
    tree.add(8);
    tree.add(7);
    tree.inorder();
    cout<<"\n";
    tree.postorder();
    cout<<"\n";
    tree.preorder();
    cout<<"\n";
    tree.DeleteHalfNodes();
    cout<<"AFTER DELETING\n";
    tree.inorder();
    cout<<"\n";
    tree.postorder();
    cout<<"\n";
    tree.preorder();
}
